contracts/Staking.sol

```diff
@@ -113,148 +96,181 @@ contract Staking is Autopetrified, ERCStaking, ERCStakingHistory, IStakingLockin
     }
 
     /**
-     * @notice Lock `_amount` staked tokens and assign `_manager` as manager with `_data` as data, so they can not be unstaked
+     * @notice Lock `_amount` staked tokens and assign `_lockManager` as manager with `_data` as data, so they can not be unstaked
      * @param _amount The amount of tokens to be locked
-     * @param _manager The manager entity for this particular lock. This entity will have full control over the lock, in particular will be able to unlock it
+     * @param _lockManager The manager entity for this particular lock. This entity will have full control over the lock, in particular will be able to unlock it
      * @param _data Data to parametrize logic for the lock to be enforced by the manager
-     * @return The id of the newly created lock
      */
-    function lock(uint256 _amount, address _manager, bytes _data) external isInitialized returns (uint256) {
+    function allowManagerAndLock(uint256 _amount, address _lockManager, uint256 _allowance, bytes _data) external isInitialized {
         Account storage account = accounts[msg.sender];
         // locking 0 tokens is invalid
         require(_amount > 0, ERROR_AMOUNT_ZERO);
+        require(_allowance >= _amount, ERROR_NOT_ENOUGH_ALLOWANCE);
 
         // check enough unlocked tokens are available
-        require(_amount <= unlockedBalanceOf(msg.sender), ERROR_NOT_ENOUGH_BALANCE);
-
-        // check not too many locks
-        require(account.activeLockIds.length < MAX_LOCKS, ERROR_TOO_MANY_LOCKS);
+        require(_amount <= _unlockedBalanceOf(msg.sender), ERROR_NOT_ENOUGH_BALANCE);
 
-        // first valid lock starts at 1, so _toLockId = 0 means no lock
-        account.lastLockId++;
-        uint256 _lockId = account.lastLockId;
-        account.activeLockIds.push(_lockId);
-        Lock storage lock_ = account.locks[_lockId];
+        Lock storage lock_ = account.locks[_lockManager];
+        // check if lock exists
+        require(lock_.allowance == 0, ERROR_LOCK_ALREADY_EXISTS);
         lock_.amount = _amount;
```

Comment:
🚨I'm thinking about the next case:
1. A user allows a lock manager with an allowance
2. The lock manager locks an amount
3. The user reduces the lock manager allowance 
4. The user calls this method

I think we will be overwriting the locked amount, I think we should add here to be safe

Reply:
In 3, it should be impossible to reduce the allowance below the locked amount.